nifi-nar-bundles/nifi-standard-services/nifi-record-sink-service-bundle/nifi-record-sink-service/src/main/java/org/apache/nifi/record/sink/LoggingRecordSink.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.controller.ControllerServiceInitializationContext;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.logging.LogLevel;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.schema.access.SchemaNotFoundException;
+import org.apache.nifi.serialization.RecordSetWriter;
+import org.apache.nifi.serialization.RecordSetWriterFactory;
+import org.apache.nifi.serialization.WriteResult;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordSet;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+@Tags({"record", "sink", "log"})
+@CapabilityDescription("Provides a RecordSinkService that can be used to log records to the application log (nifi-app.log, e.g.) using the specified writer for formatting.")
+public class LoggingRecordSink extends AbstractControllerService implements RecordSinkService {
+
+    private List<PropertyDescriptor> properties;
+    private volatile RecordSetWriterFactory writerFactory;
+    private volatile LogLevel logLevel;
+
+    public static final PropertyDescriptor LOG_LEVEL = new PropertyDescriptor.Builder()
+            .name("logsink-log-level")
+            .displayName("Log Level")
+            .required(true)
+            .description("The Log Level at which to log records (INFO, DEBUG, e.g.)")
+            .allowableValues(LogLevel.values())
+            .defaultValue(LogLevel.INFO.name())
+            .build();
+
+    @Override
+    protected void init(final ControllerServiceInitializationContext context) throws InitializationException {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(RecordSinkService.RECORD_WRITER_FACTORY);
+        properties.add(LOG_LEVEL);
+
+        this.properties = Collections.unmodifiableList(properties);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException {
+        writerFactory = context.getProperty(RecordSinkService.RECORD_WRITER_FACTORY).asControllerService(RecordSetWriterFactory.class);
+        logLevel = LogLevel.valueOf(context.getProperty(LOG_LEVEL).getValue());
+    }
+
+    @Override
+    public WriteResult sendData(RecordSet recordSet, Map<String, String> attributes, boolean sendZeroResults) throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        WriteResult writeResult;
+        try {
+            ComponentLog log = getLogger();
+            final RecordSetWriter writer = writerFactory.createWriter(getLogger(), recordSet.getSchema(), baos, attributes);
```

Comment:
Better to create the Writer in a try-with-resources block so that it always gets closed even after exception